src/System.Linq/src/System/Linq/Enumerable.cs

```diff
@@ -1041,7 +1051,13 @@ public static List<TSource> ToList<TSource>(this IEnumerable<TSource> source)
             if (source == null) throw Error.ArgumentNull("source");
             if (keySelector == null) throw Error.ArgumentNull("keySelector");
             if (elementSelector == null) throw Error.ArgumentNull("elementSelector");
-            Dictionary<TKey, TElement> d = new Dictionary<TKey, TElement>(comparer);
+
+            int capacity = 0;
+
+            if (source is TSource[]) capacity = (source as TSource[]).Length; // Cast to Array is much faster than cast to ICollection
+            else if (source is ICollection<TSource>) capacity = (source as ICollection<TSource>).Count;
```

Comment:
Good idea. 

As disscussed above, it is better not to cast to an array. But I think, the following strategy can be used here:

```C#
int capacity = 0;
var collection = source as ICollection<TSource>;
if (collection != null)
{
    capacity = collection.Count;
    if (capacity == 0)
        return new Dictionary<TKey, TElement>(comparer);
    // threshold to avoid overhead from additional cast on small collections
    if (capacity > 4) 
    {
        if (collection is TSource[])
            return ToDictionaryOnArray(collection as TSource[], keySelector, elementSelector, comparer);
        if (collection is List<TSource>)
            return ToDictionaryOnList(collection as List<TSource>, keySelector, elementSelector, comparer);
    }
}

Dictionary<TKey, TElement> d = new Dictionary<TKey, TElement>(capacity, comparer);
foreach (TSource element in source) d.Add(keySelector(element), elementSelector(element));
return d;
```

@stephentoub, what do you think?